### DISCUSSION TITLE
STM bootloader

### DISCUSSION THREAD

**Discussion by gitsr-sys** [bitcraze/discussions#429] [Q&A]
Hi,

I am trying to flash stm bootloader on Crazyflie 2.0 by using debug adapter kit and ST-Link.
I followed the instruction as described in [https://forum.bitcraze.io/viewtopic.php?t=2806&start=20](https://forum.bitcraze.io/viewtopic.php?t=2806&start=20), but I have the following result when I ran `~/crazyflie2-stm-bootloader$ make flash`

```
$ make flash
openocd -d2 -f interface/stlink-v2.cfg  -f target/stm32f4x_stlink.cfg -c init -c targets -c "reset halt" \
                 -c "flash write_image erase cf2loader.elf" -c "verify_image cf2loader.elf" -c "reset run" -c shutdown
Open On-Chip Debugger 0.10.0
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
debug_level: 2
WARNING: target/stm32f4x_stlink.cfg is deprecated, please switch to target/stm32f4x.cfg
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
adapter speed: 2000 kHz
adapter_nsrst_delay: 100
none separate
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : clock speed 1800 kHz
Info : STLINK v2 JTAG v29 API v2 SWIM v7 VID 0x0483 PID 0x3748
Info : using stlink api v2
Info : Target voltage: 3.070702
Error: init mode failed (unable to connect to the target)
in procedure 'init' 
in procedure 'ocd_bouncer'

make: *** [Makefile:170: flash] Error 1
```

Do you have any idear, @ataffanel ?

**Comment by gitsr-sys** [comment]
For your information, when I connect the cf2 with debug adapter kit, it powered up then M1 red LED flashed at about 0.5 Hz.
But when I ran `$make flash`, the cf2 rebooted.

Also,
```
$ openocd -v
Open On-Chip Debugger 0.10.0
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
```

**Comment by ataffanel** [comment]
I am not sure what can be happening and the fact that the debug probe sees 3.0V VCC is a sign that it is well connected. Just to make sure: the STM32 is connected to the debug port on the back of the Crazyflie, under the USB port, not the one on the side.

One possibility is that the stm32 is somehow kept in reset by the nRF51. Make sure the M2 LED is lit which means that the nRF51 is ON and connect USB power to the platform to make sure the power is stable.

**Comment by gitsr-sys** [comment]
@ataffanel Thank you for your answer.
I got the wrong port... Now, I try it with the port under the USB port, and got the following message;

```
~/crazyflie2-stm-bootloader$ make flash
openocd -d2 -f interface/stlink-v2.cfg  -f target/stm32f4x.cfg -c init -c targets -c "reset halt" \
                 -c "flash write_image erase cf2loader.elf" -c "verify_image cf2loader.elf" -c "reset run" -c shutdown
Open On-Chip Debugger 0.10.0
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
debug_level: 2
Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
adapter speed: 2000 kHz
adapter_nsrst_delay: 100
none separate
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : clock speed 1800 kHz
Info : STLINK v2 JTAG v29 API v2 SWIM v7 VID 0x0483 PID 0x3748
Info : using stlink api v2
Info : Target voltage: 3.091892
Info : stm32f4x.cpu: hardware has 6 breakpoints, 4 watchpoints
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       running
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
adapter speed: 1800 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x080002bc msp: 0x20005000
auto erase enabled
Info : device id = 0x10076413
Info : flash size = 1024kbytes
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x20000046 msp: 0x20005000
wrote 16384 bytes from file cf2loader.elf in 0.655472s (24.410 KiB/s)
target halted due to breakpoint, current mode: Thread 
xPSR: 0x61000000 pc: 0x2000002e msp: 0x20005000
verified 7252 bytes in 0.123996s (57.115 KiB/s)
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
adapter speed: 1800 kHz
shutdown command invoked
```

Do you think this is good to go?
After this, I tried to get the DFU mode, but I cannot still see
`Bus XXX Device XXX: ID 0483:df11 STMicroelectronics STM Device in DFU Mode` with `lsusb`.

**Reply by ataffanel** [reply]
DFU is not related to the bootloader, it is implemented in a ROM in the stm32f4 and can be activated from the nRF: make sure the Crazyflie is off, press the power switch and keep is pressed until the blue LED blinks fast (it will blink slow first and then fast). DFU should not be enabled and visible from the computer. If not, you can click the power button once more (the led will then blink even faster).

**Reply by gitsr-sys** [reply]
@ataffanel 
I see.
But when I get the correct blue LED blinks for DFU, I was not able to see any usb connection after `$ lsusb`
When I clicked the power button once more, the led blinked faster.
How can I factory reset this?
I thought I was able to flash the crazyflies firmware through DFU mode.

**Reply by tobbeanton** [reply]
Do you have a battery connected when trying it? This helps. Did you try several times?

**Reply by gitsr-sys** [reply]
@tobbeanton Hi, sorry for late response. Do you mean pushing the button and connecting the battery instead of USB cable?
Do you mean then connect USB after all this?

**Reply by ataffanel** [reply]
DFU mode can be accessed by pressing the ON/OFF switch for a very long time at startup following the procedure of my previous message. You may need to press the button once more to see the stm32 appear on the computer side).

There is not factory reset possible for the STM DFU bootloader: this is a ROM in the microcontroller we cannot erase or modify.

DFU can only program the STM32 and we consider it a recovery-mode only. The preferred ways of programming the Crazyflie is using the radio bootloader or a SWD/JTAG probe over the programming ports.

**Reply by gitsr-sys** [reply]
I see.

Only thing that I would like to have is ` Bus XXX Device XXX: ID 0483:df11 STMicroelectronics STM Device in DFU Mode` with `lsusb` for now.
Then I want to flash crazyflie2-stm-bootloader..

Even though I have correct LED flashing for the DFU mode, I cannot see any usb connection.

Any idea?

**Reply by ataffanel** [reply]
I just tested and after pressing one more time the button to get the faster blink my Crazyflie appears as DFU on usb (I get it on lsusb). If this is not the case for you it is likely a hardware problem with your particular Crazyflie.
